mlflow/gateway/providers/mosaicml.py

```diff
@@ -0,0 +1,226 @@
+            payload=payload,
+        )
+
+
+    def _parseChatMessagesToPrompt(self, messages: List[chat.RequestMessage]) -> str:
+        """
+        This parser is based on the format described in
+        https://huggingface.co/blog/llama2#how-to-prompt-llama-2 .
+        The expected format is:
+            "<s>[INST] <<SYS>>
+            {{ system_prompt }}
+            <</SYS>>
+
+            {{ user_msg_1 }} [/INST] {{ model_answer_1 }} </s><s>[INST] {{ user_msg_2 }} [/INST]"
+        """
+        if all(m1 != m2 for m1, m2 in zip(messages, messages[1:])):
+            raise MlflowException.invalid_parameter_value(
+                "Consecutive messages cannot have the same 'role'.",
+            )
+        prompt = ""
+        for m in messages:
+            if m.role == "system":
+                prompt += f"<s>[INST] <<SYS>> ${m.content} <</SYS>>"
+            elif m.role == "user":
+                inst = f" ${m.content} [/INST]"
+                if prompt.endswith("<</SYS>>"):
+                    inst = "<s>[INST]" + inst
+                prompt += inst
+            elif m.role == "assistant":
+                if not prompt.endswith("[/INST]"):
+                    raise MlflowException.invalid_parameter_value(
+                        "Messages with role 'assistant' must be preceeded by a message "
+                        "with role 'user'."
+                    )
+                prompt += f" ${m.content} </s>"
+        return prompt
+
+    async def chat(self, payload: chat.RequestPayload) -> chat.ResponsePayload:
+        payload = jsonable_encoder(payload, exclude_none=True)
+        self.check_for_model_field(payload)
+        key_mapping = {
+            "max_tokens": "max_new_tokens",
+        }
+        for k1, k2 in key_mapping.items():
+            if k2 in payload:
+                raise HTTPException(
+                    status_code=400, detail=f"Invalid parameter {k2}. Use {k1} instead."
+                )
+        payload = rename_payload_keys(payload, key_mapping)
+
+        # Handle 'prompt' field in payload
+        prompt = [self._parseChatMessagesToPrompt(payload.pop("messages"))]
+
+        # Construct final payload structure
+        final_payload = {"inputs": prompt, "parameters": payload}
+
+        # Input data structure for Mosaic Text Completion endpoint
+        #
+        # {"inputs": [prompt],
+        #  {
+        #    "parameters": {
+        #      "temperature": 0.2
+        #    }
+        #   }
+        # }
+
+        resp = await self._request(
+            self.config.model.name,
+            final_payload,
+        )
+        # Response example (https://docs.mosaicml.com/en/latest/inference.html#text-completion-models)
+        # ```
+        # {
+        #   "outputs": [
+        #     "string",
+        #   ],
+        # }
+        # ```
+        return chat.ResponsePayload(
+            **{
+                "candidates": [
+                    {
+                        "message": {
```

Comment:
I think we need to map the full response back as well?
Isn't the model going to return something like this:

```
 "<s>[INST] <<SYS>>
            {{ system_prompt }}
            <</SYS>>
            {{ user_msg_1 }} [/INST] {{ model_answer_1 }} </s>"
```

W/ a `model_answer_1` added to it? Assistant response would be the `model_answer_1`, but we should return the previous `system_prompt` and `user_msg_1` as well?  This way, from the user perspective, we get the full history of messages back and forth.

We can test this out against staging to verify the behavior.

Reply:
From: 
https://huggingface.co/blog/llama2#how-to-prompt-llama-2

"The model is stateless and does not "remember" previous fragments of the conversation, we must always supply it with all the context so the conversation can continue. This is the reason why context length is a very important parameter to maximize, as it allows for longer conversations and larger amounts of information to be used."

Reply:
Nope, it just returns output. Tested it out here: https://huggingface.co/blog/llama2#how-to-prompt-llama-2

Reply:
Ah, ok, I see. Sounds like we functionality for gateway in the future :-)